mlflow/utils/model_utils.py

```diff
@@ -428,3 +429,58 @@ def _validate_pyfunc_model_config(model_config):
             "valid file path or of type ``dict`` with string keys.",
             error_code=INVALID_PARAMETER_VALUE,
         )
+
+
+RECORD_ENV_VAR_ALLOWLIST = {
+    # api key related
+    "API_KEY",
+    "API_TOKEN",
+    # databricks auth related
+    "DATABRICKS_HOST",
+    "DATABRICKS_USERNAME",
+    "DATABRICKS_PASSWORD",
+    "DATABRICKS_TOKEN",
+    "DATABRICKS_INSECURE",
+    "DATABRICKS_CLIENT_ID",
+    "DATABRICKS_CLIENT_SECRET",
+    "_DATABRICKS_WORKSPACE_HOST",
+    "_DATABRICKS_WORKSPACE_ID",
+    # dbconnect related
+    "SPARK_REMOTE",
+    "SPARK_LOCAL_REMOTE",
+}
+
+
+@contextlib.contextmanager
+def env_var_tracker():
+    """
+    Context manager for temporarily tracking environment variables accessed.
+    It tracks environment variables accessed during the context manager's lifetime.
+    """
+    from mlflow.environment_variables import MLFLOW_RECORD_ENV_VARS_IN_MODEL_LOGGING
+
+    tracked_env_names = set()
+
+    if MLFLOW_RECORD_ENV_VARS_IN_MODEL_LOGGING.get():
+        original_getitem = os._Environ.__getitem__
+        original_get = os._Environ.get
```

Comment:
Is there any chance that `__getitem__` is not the original `__getitem__`? What happens if we have two threads and this context manager runs in both of them?

Reply:
I would think of this as an edge case, would people invoke logging model in multiple threads?

Reply:
> would people invoke logging model in multiple threads?

I think yes. What would happen if `__getitem__` is the patched `__getitem__`? Would that cause infinite recursion?

Reply:
Can you run this?

```python
from mlflow.utils.model_utils import env_var_tracker
import os


with env_var_tracker():
    with env_var_tracker():
        print(os.environ.get("FOO"))

```

Reply:
```
with env_var_tracker():
...     with env_var_tracker():
...             print(os.environ.get("FOO"))
...             print(os.environ["HOME"])
... 
None
/Users/serena.ruan
```

Reply:
Thanks for the confirmation!